fleetspeak/src/e2etesting/lib/setup_components.go

```diff
@@ -0,0 +1,318 @@
+	}
+	err = os.Mkdir(path.Join(tempPath, "textservices"), 0777)
+	if err != nil {
+		return fmt.Errorf("Unable to create textservices directory: %v", err)
+	}
+	err = os.Mkdir(path.Join(tempPath, "services"), 0777)
+	if err != nil {
+		return fmt.Errorf("Unable to create services directory: %v", err)
+	}
+	_, err = os.Create(path.Join(tempPath, "communicator.txt"))
+	if err != nil {
+		return fmt.Errorf("Unable to create communicator.txt: %v", err)
+	}
+	err = ioutil.WriteFile(path.Join(tempPath, "textservices", "frr.textproto"), []byte(proto.MarshalTextString(&clientServiceConf)), 0644)
+	if err != nil {
+		return fmt.Errorf("Unable to write frr.textproto file: %v", err)
+	}
+
+	// Server services configuration
+	serverServiceConf := servicesPb.ServiceConfig{Name: "FRR_server", Factory: "GRPC"}
+	grpcConfig := grpcServicePb.Config{Target: fmt.Sprintf("localhost:%v", fsFrontendPort), Insecure: true}
+	serverServiceConf.Config, err = ptypes.MarshalAny(&grpcConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal grpcConfig: %v", err)
+	}
+	serverConf := servicesPb.ServerConfig{Services: []*servicesPb.ServiceConfig{&serverServiceConf}, BroadcastPollTime: new(duration.Duration)}
+	serverConf.BroadcastPollTime.Seconds = 1
+	builtServerServicesConfigPath := path.Join(tempPath, "server.services.config")
+	err = ioutil.WriteFile(builtServerServicesConfigPath, []byte(proto.MarshalTextString(&serverConf)), 0644)
+	if err != nil {
+		return fmt.Errorf("Unable to write server.services.config: %v", err)
+	}
+	return nil
+}
+
+func (cp *componentPids) start(tempPath string, fsFrontendPort, fsAdminPort int) error {
+	// Start server
+	serverRunCmd := exec.Command("server", "-logtostderr", "-components_config", path.Join(tempPath, "server.config"), "-services_config", path.Join(tempPath, "server.services.config"))
+	serverPidChan := make(chan int, 1)
+	go func() {
+		cp.serverPid = <-serverPidChan
+	}()
+	go startProcess(serverRunCmd, serverPidChan)
+
+	serverStartTime := time.Now()
+
+	// Start client
+	clientRunCmd := exec.Command("client", "-logtostderr", "-config", path.Join(tempPath, "linux_client.config"))
+	clientPidChan := make(chan int, 1)
+	go func() {
+		cp.clientPid = <-clientPidChan
+	}()
+	go startProcess(clientRunCmd, clientPidChan)
+
+	// Get new client's id and start service in current process
+	clientID, err := waitForNewClientID(fsAdminPort, serverStartTime)
+	if err != nil {
+		return fmt.Errorf("No new clients have connected: %v", err)
+	}
+	serviceRunCmd := exec.Command(
+		"python",
+		"frr_python/frr_server.py",
+		fmt.Sprintf("--client_id=%v", clientID),
+		fmt.Sprintf("--fleetspeak_message_listen_address=localhost:%v", fsFrontendPort),
+		fmt.Sprintf("--fleetspeak_server=localhost:%v", fsAdminPort))
+	servicePidChan := make(chan int, 1)
+	go func() {
+		cp.servicePid = <-servicePidChan
+	}()
+	startProcess(serviceRunCmd, servicePidChan)
+	return nil
+}
+
+func main() {
```

Comment:
Further simplyfing the printing of errors:

func run() error {
  // contents of main, but returning errors instead of printing
}

func main() {
  err := run()
  if err != nil {
     // print error and exit
  }
}